backend/layers/persistence/persistence.py

```diff
@@ -507,6 +507,46 @@ def get_all_datasets(self) -> Iterable[DatasetVersion]:
             dataset_version_ids.extend(collection.datasets)
         return self._get_datasets(dataset_version_ids)
 
+    def get_all_published_collections_with_datasets(self) -> List[CollectionVersionWithDatasets]:
+        """
+        Returns all published collections with their published datasets
+        """
+        active_collections = self.get_all_mapped_collection_versions()
+
+        dataset_version_ids = []
+        active_collections_list = list(active_collections)
+        for collection in active_collections_list:
+            dataset_version_ids.extend(collection.datasets)
+        active_datasets = self._get_datasets(dataset_version_ids)
+        datasets_by_collection_id = {}
+        for dataset in active_datasets:
+            if collection_datasets := datasets_by_collection_id.get(dataset.collection_id.id):
```

Comment:
I think we can simplify this loop block by initializing datasets_by_collection_id as a defaultdict(list) and then doing:
```
for dataset in active_datasets:
      datasets_by_collection_id[dataset.collection_id.id].append(dataset)
```